Initial

wojtekmach opened the pull request.

wojtekmach commented on config/config.exs:
```diff
@@ -0,0 +1,94 @@
+#     # Mirrors are read-only and must be configured with upstream URL and public key.
+#     #
+#     # The optional `:only` option configures an allowed packages list to fetch.
+#     # When using `:only` option, you need to manually make sure that all of package's
```
I added the `:only` option mostly for testing this locally, but I can also see people using it for poor man's curation of packages.

What do you think about this option? It's perhaps kinda weird that on one hand repository says in `/names` that it has a bunch of packages, but it might effectively have just a small subset. An alternative is to have mirror maintain it's own resource files that match the actual packages. In that case, however, the mirror would have to sign the resources with it's own private key, so the mirror would no longer work out of the box with Hex.pm public key that ships with the client.

ericmj replied:
I think having a partial mirror of only packages you use is a very good use case. Will the mirror re-sign resources, if that's the case it can build its own `names` and `versions`?

wojtekmach replied:
My plan is to ship with a partial mirror like above which *does not* re-sign resources (and have caveats mentioned above), but folks could write an alternate mirror that does re-sign. WDYT?

ericmj replied:
We have three use-cases for nimble_hex as I see it:

  1. Run it only as a mirror of hexpm
  2. Run it only as its own repository
  3. Do both of the above

For 1. we don't have to re-sign. For 2. we need to sign our own resources. How will 3. work though? Will it run two distinct repositories, one for the mirror and one for its own – or is the mirror merged into the same repository?

wojtekmach replied:
Oh, I should improve documentation around this, nimble_hex supports multiple repositories out of the box, the URLs always have a repo in the path:

```
GET /repos/hexpm_mirror/names
GET /repos/my_repo/names
GET /repos/hexpm_mirror/names
POST /api/repos/my_repo/publish
```

btw, I was thinking of introducing some default repo config so that a plug can rewrite the `GET /names` request into `GET /repos/<default_repo>/names` but we don't need it yet.

ericmj replied:
Perfect! Then don't re-sign and add the caveat about names and versions imo.

wojtekmach commented on lib/nimble_hex/mirror/server.ex:
```diff
@@ -0,0 +1,151 @@
+  end
+
+  defp sync(mirror) do
+    # TODO: currently sync is naive and downloads everything everytime.
```
will finish this up once we figure out what to do with the `:only` option.

wojtekmach commented on lib/nimble_hex/mirror/server.ex:
```diff
@@ -0,0 +1,151 @@
+        sync_release(mirror, config, name, version)
+      end
+
+      # TODO: sync docs?
```
:eyes:

ericmj replied:
I would say no. nimble_hex mirrors the repository and docs are not in the repo

wojtekmach replied:
That's a good point, agreed.

FWIW even though it's kind of implementation detail of hex.pm that there happens to be `repo.hex.pm/docs/<package>-<version>.tar.gz`, I'd consider lifting that to the spec and implement appropriate APIs in hex_core. We kinda set a precedent for adding something like this by considering https://github.com/hexpm/hex_core/issues/61.

Anyhow, I'm fine with leaving this out for now, will start a discussion about this elsewhere.

wojtekmach commented on lib/nimble_hex/router.ex:
```diff
@@ -0,0 +1,75 @@
+    send_resp(conn, 404, "not found")
+  end
+
+  # TODO: pass allowed repos to router and only allow them here
```
👀

wojtekmach commented on README.md:
```diff
@@ -0,0 +1,150 @@
+* Support for multiple repositories
+* Mirroring
+
+## Setup
```
General feedback on the docs would be very appreciated. There's some overlap between this and comments in `config/config.exs` so any ideas how to improve it would be appreciated.

wojtekmach replied:
I also need to write a section about setting the expectations of the project, for people to actually use it they'd need to configure backups, ssl, auth, and more, and that's out of the scope of this project.

wojtekmach commented on README.md:
```diff
@@ -0,0 +1,150 @@
+See [`mix help hex.config`](https://hexdocs.pm/hex/Mix.Tasks.Hex.Config.html) for more information
+about configuring your Hex installation.
+
+## Deploying with releases
```
I feel like this section is unnecessarily complicated by the S3 setup, do you think we should split it up?  I wanted to have s3 setup committed to code so that I can easily test it locally, but at the same time I think people may not want to use it, so perhaps it's better to default release setup to use local filesystem adapter instead?

Or maybe we assemble two different releases, one with local storage backend and one with S3? That seems it may overly complicate things too, not sure.

ericmj replied:
Let's keep it simple in the releases examples and document S3 separately from the release docs?

wojtekmach commented on README.md:
```diff
@@ -0,0 +1,150 @@
+* Support for multiple repositories
+* Mirroring
+
+## Setup
```
I think it's important for this project to ship with step-by-step instructions both on setting up the server as well as configuring clients to use it. I intend to also write a similar guide to use with Rebar.

I wonder if README is the best place for such "guides", perhaps we should turn these into ExDoc pages and publish on hex docs? (without actually publishing nimble_hex as package). Or add these instructions to https://hex.pm/docs/usage?

I think we should iterate on docs in the README, but just something to think about for the future.

ericmj replied:
I don't think we should publish to hexdocs only or are you suggesting pushing nimble_hex as a package?

ericmj replied:
For now I think it's fine to put docs in the README or maybe have a separate USAGE.md document?

wojtekmach replied:
> or are you suggesting pushing nimble_hex as a package?

Initially I was considering this but for now decided against as it's adding complexity that we don't need (if we do that we'd likely make dependency on ex_aws optional, etc). Yeah, I was thinking about publishing just to hex docs but not publishing it as a package. Agreed, let's keep it simple for now. USAGE.md seems like a good idea to, will look into it.

wojtekmach commented:
cc @ericmj @supersimple @josevalim

ericmj commented on README.md:
```diff
@@ -0,0 +1,150 @@
+  defp hex() do
+    [
+      api_url: "http://localhost:4000/api/repos/test_repo",
+      api_key: "does-not-matter"
```
Does `api_key` have to be set? I am guessing this is so that the key for hex.pm isn't sent to other servers?

wojtekmach replied:
oh, I was meaning to add a comment for it, will do that because it's confusing.

Without this, when the user is not authed already, with mix/hex they'd get an auth prompt which will try to hit an http api endpoint that isn't implemented. And when the user happened to be already authed (e.g. against hex.pm/api), mix/hex would use that authed key but it's never checked as nimble_hex http api does not do any auth.

That's a good point that with this we'd additionally not leak api key to other servers.

ericmj commented on config/config.exs:
```diff
@@ -0,0 +1,94 @@
+#     ]
+#   ]
+
+import_config "#{Mix.env()}.exs"
```
This file seems mostly duplicated with `config/releases.exs`, let's leave it blank and move the explanatory comments to `config/releases.exs`?

ericmj commented on lib/nimble_hex/repository.ex:
```diff
@@ -0,0 +1,7 @@
+defmodule NimbleHex.Repository do
+  @moduledoc false
+
+  @derive {Inspect, only: [:name, :public_key, :store, :registry]}
```
Nice!

wojtekmach commented:
@ericmj updated! I've moved docs around, now the README has an overview of features and step-by-step guide of using it with Hex and Rebar3 (todo). I've moved more comprehensive docs to module docs (which are linked to at the end of the README)

I've also added a more realistic mirror implementation that calculates the diff and only downloads what's required. It also has the same mechanism of loading/saving registry backup as the regular repos. The mirror implementation currently doesn't handle when someone overwrites a given release, I'll add an issue for it later. 

WDYT?

ericmj commented on README.md:
```diff
@@ -0,0 +1,219 @@
+
+We welcome anyone to contribute to the project, especially around documentation and guides,
+but features specific to narrow set of users likely won't be accepted. For a complete and
+production-grade implementation see source code of [Hex.pm](https://github.com/hexpm/hexpm).
```
Should we clarify further that hexpm is not for running your organization's Hex server, rather it's for running Hex.pm specifically which has very different constraints?

wojtekmach replied:
good call. How about:

For a complete and
 production-grade implementation see source code of [Hex.pm](https://github.com/hexpm/hexpm), however keep in mind it's optimised for running the public registry for the community and thus have different features and constraints than a self-hosted solution might have.

ericmj replied:
Great!

wojtekmach commented:
I've merged it so that people can easily start testing it, but feedback on the PR is still welcome of course!
